Implement projection tooling for simple P/Invoke cases

kotlarmilos opened the pull request:
Most of the code has been moved from https://github.com/xamarin/binding-tools-for-swift. 

This PR is quite large, and there are multiple options for reviewing it. It includes all necessary components to demonstrate the end-to-end flow and allow for CI testing.

I have attempted to prune the tooling as much as possible, though there are some limitations. When reviewing this PR, I recommend initially focusing on the tooling UX and the quality of the generated bindings rather than the internal workings of the tooling. 

Here are my suggestions for the review process: Begin with the `SwiftRuntimeLibrary`, it is a small project that outlines the implementation of Swift types in C#. The `SwiftBindings` project includes the tool interface and compilers for bindings. The `SwiftReflector` project includes ABI aggregation logic, module declarations, and type system. The `SyntaxDynamo` project includes API for C# source code generation and can potentially be improved at a later stage as it shouldn't impact the quality of the generated bindings.

## Description

This PR represents a minimal set of features required for a simple direct P/Invoke scenario. It includes the following components, along with unit tests and samples.
- **SwiftBindings**: Command-line interface that orchestrates the tooling workflow.
- **SwiftReflector**: A library provides support for public ABI aggregation. It contains implementation of module declarations and the type system.
- **SwiftRuntimeLibrary**: Library that provides runtime marshaling support and projections of common Swift types.
- **SyntaxDynamo**: Library that provides an API for generating C# source code.

## Usage

Options:
```
Description:
  Swift bindings generator.

Usage:
  SwiftBindings [options]

Options:
  -a, --swiftabi <swiftabi> (REQUIRED)  Path to the Swift ABI file.
  -o, --output <output> (REQUIRED)      Output directory for generated bindings.
  -v, --verbose <verbose>               Prints information about work in process.
  -h, --help                            Display a help message.
  --version                             Show version information
  -?, -h, --help                        Show help and usage information
```

Example usage:
```sh
dotnet SwiftBindings.dll -a "./HelloLibrary.abi.json" -o "./"
```
If a unsupported type is encountered, the tooling will ignore it and generate C# source code for known syntax.

## Simple binding

The supported scenario involves simple P/Invoke calls without primitive params and no exceptions.

Swift library example:
```swift
public func sayHello() {
    print("Hello world")
}
```

User's code in C#:
```csharp
using System;
using HelloLibraryBindings;

namespace HelloWorld
{
    public class Program
    {
        public static void Main(string[] args)
        {
            HelloLibrary.sayHello();
        }
    }
}
```

Generated bindings in C#:
```csharp
using System;
using System.Runtime.InteropServices;
namespace HelloLibraryBindings
{
    public class HelloLibrary
    {
        [DllImport("libHelloLibrary.dylib", EntryPoint = "$s12HelloLibrary03sayA0yyF")]
        internal static extern void PIfunc_sayHello();
        public static void sayHello()
        {
            PIfunc_sayHello();
        }
    }
}
```

The design discussion for the simple P/Invoke scenarios is available at https://github.com/dotnet/runtimelab/issues/2518. The documentation is available at https://github.com/dotnet/runtimelab/pull/2512.

kotlarmilos committed:
Add SwiftRuntimeLibrary project

The SwiftRuntimeLibrary includes C# implementation of Swift types used for marshalling. This commit adds initial layout of the project.

kotlarmilos committed:
Add SyntaxDynamo project

The SyntaxDynamo is a code generator for C# source code generation. Also, it contains Swift syntax and is used as a parser for API aggregation.

kotlarmilos committed:
Add SwiftReflector project

Introduces type mapping and API reflection of the tooling. Key components:
- Inventory: Collects mangled names and generates entry points mapping.
- Reflector and SwiftXmlReflection: Aggregates Swift public API.
- TopLevelFunctionCompiler: Provides tools to generate C# methods, properties, and P/Invoke definitions.
- TypeMapping: Facilitates Swift to C# type conversion, maintaining a type database.
- BindingsCompiler: Handles generating the C# bindings onto the types from the Swift module.

kotlarmilos committed:
Add SwiftBindings project

A command-line interface that orchestrates the tooling workflow.

kotlarmilos committed:
Add basic testing coverage

kotlarmilos committed:
Fix sample application

rolfbjarne commented on src/samples/HelloWorld/.gitignore:
```diff
@@ -0,0 +1,6 @@
+*.swiftinterface*
```
Maybe this .gitignore should go in the src/samples directory (or even higher up?)

kotlarmilos replied:
Moved to the global `.gitignore`.

rolfbjarne commented on src/SwiftRuntimeLibrary/Exceptions.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace SwiftRuntimeLibrary {
+	public static class Exceptions {
+		public static T ThrowOnNull<T>(T o, string name, string message = null) where T : class
```
There's ArgumentNullException.ThrowIfNull: https://learn.microsoft.com/en-us/dotnet/api/system.argumentnullexception.throwifnull?view=net-8.0

kotlarmilos replied:
Updated.

rolfbjarne commented on src/SwiftReflector/Demangling/Swift4Demangler.cs:
```diff
@@ -0,0 +1,1814 @@
+// Copyright (c) Microsoft Corporation.
```
Do we need a Swift 4 demangler? I though we were only going to supportd Swift 5+.

kotlarmilos replied:
Right, it is not needed and will be excluded from this PR.

jkoritzinsky commented on src/SwiftBindings/src/SwiftBindings.csproj:
```diff
@@ -2,9 +2,15 @@
+    <ProjectReference Include="../../SyntaxDynamo/SyntaxDynamo.csproj" />
+    <ProjectReference Include="../../SwiftRuntimeLibrary/SwiftRuntimeLibrary.csproj" />
 
+    <PackageReference Include="CommandLineParser" Version="2.9.1" />
```
Use `System.CommandLine` to get a consistent CLI UX with the rest of the .NET product?

kotlarmilos replied:
Thanks, updated.

jkoritzinsky commented on src/SwiftBindings/tests/CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+cmake_minimum_required(VERSION 3.24)
+
+project(SmokeTests)
```
Let's use the Swift language support in CMake and require using a generator that supports it (ie. Ninja). As this code isn't going in the runtime repo, we can be more reasonable about requiring specific features/options.

```suggestion
project(SmokeTests Swift)
```

kotlarmilos replied:
It seems that the `ninja` generator is not installed on build machines: https://github.com/actions/runner-images/blob/macOS-12/20240202.1/images/macos/macos-12-Readme.md

jkotas replied:
> https://github.com/actions/runner-images/blob/macOS-12/20240202.1/images/macos/macos-12-Readme.md

Nit: We use our build images. This link is for GitHub Actions images - I do not think it is relevant to our builds.

jkoritzinsky reviewed (commented):
Some basic comments around the sample/infra to start.

I'm a little concerned by the fact that we have a totally independent C# syntax writer here that also seems to understand Swift syntax, not just SwiftInterface syntax (likely from when BTfS would write separately). Maybe we can delete the Swift-writing support or move it into a separate assembly now that that's not on the roadmap?

stephen-hawley commented on src/SwiftReflector/SwiftInterfaceReflector/GeneratedParser/SwiftInterfaceBaseListener.cs:
Generated code shouldn't get checked in. The original BTfS repository has tooling to get the appropriate version of antlr and generate the parser as well as debugging code for it.

kotlarmilos replied:
I agree with you, it should be generated and not checked in. To generate them during the build process, Java support would be required, including the compilation of .java source files.

There is an ABI file in json format generated by the Swift compiler that could be utilized instead, see https://github.com/dotnet/runtimelab/issues/2518#issuecomment-1948482620. It should simplify the process by eliminating the dependency on Java and avoid additional files such as dynamic libraries or swiftinterface files.

kotlarmilos committed:
Remove Swift4 demangler

stephen-hawley commented on README.md:
````diff
@@ -2,9 +2,18 @@
+
+Options:
+```
+  -d, --dylibs             Required. Paths to the dynamic libraries (dylibs), separated by commas.
````
The original design was to operate on multiple libraries, but frankly, this made things worse overall. I feel like we should simplify and make it take only a single dylib/swiftinterface pair.

stephen-hawley commented on src/SwiftReflector/CSKeywords.cs:
```diff
@@ -0,0 +1,44 @@
+
+		public static bool IsKeyword (string s)
+		{
+			Exceptions.ThrowOnNull (s, "s");
```
#nullable enable was a glint in C#'s eye when I started this project. Here's a perfect opportunity to fix this and add it in and remove the ThrowOnNull calls when not needed.

kotlarmilos replied:
Thanks, added to the task list.

stephen-hawley commented on src/SwiftReflector/Demangling/Decomposer.cs:
```diff
@@ -0,0 +1,1175 @@
+		const string kSwiftID = "__T";
+		public const string kSwift4ID = "__T0";
+		public const string kSwift4xID = "_$s";
+		public const string kSwift5ID = "_$S";
```
As we discussed in the meeting, I think it's reasonable to shed everything before swift 5 and throw if we get anything else.

kotlarmilos replied:
Thanks, removed. Is it safe to remove the `kSwift4xID`? It is used in
```csharp
f (swiftIdentifier.StartsWith (kSwift4xID, StringComparison.Ordinal)|| swiftIdentifier.StartsWith (kSwift5ID, StringComparison.Ordinal)) {
```

stephen-hawley commented on src/SwiftReflector/Demangling/Swift4NodeToTLDefinition.cs:
Feel free to exclude swift 4 things.

stephen-hawley commented on src/SwiftReflector/Extensions.cs:
```diff
@@ -0,0 +1,238 @@
+
+		public static bool IsSwiftEntryPoint (this NListEntry entry)
+		{
+			return !String.IsNullOrEmpty (entry.str) && (entry.str.StartsWith ("__T", StringComparison.Ordinal) ||
```
`__T` and `_$s` are legacy

kotlarmilos replied:
According to the ABI document, the `_$s` is used for globals: https://github.com/apple/swift/blob/main/docs/ABI/Mangling.rst#L11

Did you mean `_$S` perhaps?

kotlarmilos committed:
Update the BindingsTool class to use System.CommandLine

kotlarmilos committed:
Remove SyntaxDynamo.SwiftLang files

kotlarmilos committed:
Update README.md

kotlarmilos committed:
Remove Swift4 support

kotlarmilos committed:
Rename verbose argument

kotlarmilos committed:
Update src/SwiftBindings/tests/CMakeLists.txt

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

kotlarmilos committed:
Use ninja generator for cmake

kotlarmilos committed:
Fix lint formatting

kotlarmilos committed:
Update bindings compiler to use a module name and emit pinvokes and static functions within the same class

kotlarmilos committed:
Update Microsoft.DotNet.Arcade.Sdk version

kotlarmilos committed:
Add .sln file

kotlarmilos committed:
Add nuget.org as package source

jkotas commented:
> Add nuget.org as package source

Why do you need this?

It violates the security rule. I expect that the official builds of the branch are going to fail because of this.

kotlarmilos committed:
Remove ninja generator to test the CI

kotlarmilos commented:
> > Add nuget.org as package source
> 
> Why do you need this?
> 
> It violates the security rule. I expect that the official builds of the branch are going to fail because of this.

I am trying to setup the CI tests and the tooling references the `Antlr4.Runtime.Standard` library for creating the `.swiftinterface` parser. Thanks for the heads up, I will revert the change.

kotlarmilos committed:
Revert temporary changes

jkotas commented:
https://github.com/dotnet/arcade/blob/main/Documentation/MirroringPackages.md is the way to do it correctly.

kotlarmilos committed:
Update .gitignore

kotlarmilos committed:
Use cmake instead of ninja

kotlarmilos committed:
Use macOS vmImage

kotlarmilos committed:
Don't override SWIFT_PLATFORM_SUFFIX if set

kotlarmilos committed:
Fix SWIFT_COMPILER_TARGET variable

kotlarmilos committed:
Use cmake instead of ninja

kotlarmilos committed:
Introduce Swift types database

matouskozak commented on src/SwiftReflector/MachOHawley.cs:
What is the reason behind using the "MachOHawley.cs" file name? Is it a reference to @stephen-hawley or does it have a connection with Swift/Apple? Sorry, I'm not that well oriented in this domain but it caught my eye :)

kotlarmilos replied:
I believe it is not Swift-specific, but Steve can say for sure :)

stephen-hawley replied:
> I believe it is not Swift-specific, but Steve can say for sure :)

It's because MachO.cs was taken from macios, but when Swift was introduced, Apple added some things that (1) we cared about and (2) broke MachO.cs and (3) macios didn't need/want (yet). It can be renamed. It's not important.

kotlarmilos commented:
> Maybe we can delete the Swift-writing support or move it into a separate assembly now that that's not on the roadmap?

I agree, let's try to narrow the scope at this point.

> Generated code shouldn't get checked in. The original BTfS repository has tooling to get the appropriate version of antlr and generate the parser as well as debugging code for it.

I've implemented (and will upstream) a prototype ABI parser that consumes `.abi.json` instead of `.swiftinterface` and `.dylib`. The prototype creates module declarations and relies on the underlying type system. It removes dependencies on `Demangling`, `Inventory`, `SwiftInterfaceParser`, and `Antlr4.Runtime.Standard` (Java). Additionally, it should simplify the UX.

kotlarmilos committed:
Temporarily remove SwiftInterfaceReflector from the project

This change removes SwiftInterfaceReflector and its dependencies to simplify code review. Temporarily, the tooling will consume the ABI interface instead of swiftinterface. The code has been reorganized into components: SwiftBindings, which generates C# bindings, and SwiftReflector, which contains type system ABI aggregation logic. The ISwiftParser interface has been introduced to facilitate modularization within the project, and ABI aggregators are expected to implement this interface.

kotlarmilos committed:
Refactor null exceptions to use ArgumentNullException.ThrowIfNull

kotlarmilos committed:
Update target framework to net8.0

kotlarmilos committed:
Add src and tests directories for projects

kotlarmilos committed:
Introduce PInvoke tests

kotlarmilos committed:
Add uniqueId to CompiledAssembly name

kotlarmilos committed:
Update runtimelab.yml pipeline configuration

kotlarmilos committed:
Change osGroup from Linux to OSX in runtimelab.yml

kotlarmilos committed:
Add sample bindings to .gitignore

rolfbjarne reviewed (commented):
> If a unsupported type is encountered, the tooling will ignore it and generate C# source code for known syntax.

So one test scenario would be to throw all the .swiftinterface files Xcode ships at it and verify that the binding process doesn't fail.

```shell
for sw in $(find /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks -name '*.swiftinterface'); do
    dotnet SwiftBindings.dll --swiftinterface $sw ...
done
```

kotlarmilos commented:
The scope of the tooling has been narrowed to only support projecting public functions into static C# methods with P/Invoke declarations. The tooling has been split into two phases: public ABI aggregation and code generation. The integration point is ISwiftParser and ModuleDeclaration as implemented in https://github.com/dotnet/runtimelab/blob/swift-bindings/basic-tooling/src/SwiftBindings/src/Program.cs#L69-L86.

For simplicity, the Swift interface parser has been temporarily removed and will be introduced in subsequent PRs. Stress tests have been added to https://github.com/dotnet/runtimelab/tree/swift-bindings/basic-tooling/src/SwiftBindings/tests (kudos to @jakobbotsch).

Please reflect on important points that you think should be addressed in this PR, with focus on the `SwiftBindings` project.

kotlarmilos commented:
> So one test scenario would be to throw all the .swiftinterface files Xcode ships at it and verify that the binding process doesn't fail.

Good idea. I suggest to bundle it with `.swiftinterface` parser changes.

kotlarmilos commented:
To facilitate easier code review, this PR will be divided into smaller PRs:
 - [ ] https://github.com/dotnet/runtimelab/pull/2525 ~that provides runtime marshaling support and projections of common Swift types~ provides an MVP implementation of bindings generator with simplified parser and emitter.
 ~- [ ] https://github.com/dotnet/runtimelab/pull/2526 that provides an API for generating C# source code~
 ~- [ ] SwiftReflector project that provides support for public ABI aggregation~
 ~- [ ] SwiftBindings project that provides a command-line interface~

kotlarmilos commented:
After discussion in https://github.com/dotnet/runtimelab/pull/2526, we are closing this PR in favor of an MVP proposed in https://github.com/dotnet/runtimelab/pull/2525.
